File: Java/Todo_app/src/main/java/com/todo/service/AuthService.java
package com.todo.service;

import com.todo.dao.UserDao;
import com.todo.model.User;

import java.sql.SQLException;
import java.util.Optional;

public class AuthService {
    private final UserDao userDao = new UserDao();
    private User currentUser;


    // validation for registering new user
    public User register(String username, String password) throws SQLException{
        if (username == null || username.trim().isEmpty()) {
            throw new IllegalArgumentException("Username cannot be empty.");
        }
        if (password == null || password.trim().isEmpty()) {
            throw new IllegalArgumentException("Password cannot be empty.");
        }

        String normalizedUsername = username.trim();

        Optional<User> existing = userDao.findByName(normalizedUsername);
        if (existing.isPresent()) {
            throw new IllegalArgumentException("User already exists, please select a different username.");
        }

        currentUser = userDao.createUser(normalizedUsername, password);
        return currentUser;
    }

    //validation for user login
    public User login(String username, String password) throws SQLException {
        if (username == null || username.trim().isEmpty()) {
            throw new IllegalArgumentException("Invalid username or password");
        }
        if (password == null || password.trim().isEmpty()) {
            throw new IllegalArgumentException("Invalid username or password");
        }

        Optional<User> user = userDao.validateLogin(username.trim(), password);
        if (user.isEmpty()) {
            throw new IllegalArgumentException("Invalid username or password");
        }

        currentUser = user.get();
        return currentUser;
    }

    public void logout() {
        currentUser = null;
    }

    public boolean isLoggedIn() {
        return currentUser != null;
    }

    public User getCurrentUser() {
        return currentUser;
    }
}

File: JavaScript/src/services/AuthService.js
/**
 * Authentication service for user registration and login
 * Validates user input and enforces unique username constraint
 * Handles credential verification for secure access
 */
import { UserDAO } from '../dao/UserDAO.js';

export class AuthService {
  constructor() {
    this.userDAO = new UserDAO();
  }

  /**
   * Register new user with validation
   * Checks for empty fields and duplicate usernames
   * @param {string} name - Username
   * @param {string} password - User password
   * @returns {Promise<User>} Created user object
   */
  async register(name, password) {
    if (!name || name.trim() === '') {
      throw new Error('Username cannot be empty');
    }
    if (!password || password.trim() === '') {
      throw new Error('Password cannot be empty');
    }
    const existingUser = await this.userDAO.findByName(name);
    if (existingUser) {
      throw new Error('Username already exists');
    }
    return await this.userDAO.createUser(name, password);
  }

  /**
   * Authenticate user credentials
   * Verifies username exists and password matches
   * @param {string} name - Username
   * @param {string} password - User password
   * @returns {Promise<User>} Authenticated user object
   */
  async login(name, password) {
    const user = await this.userDAO.findByName(name);
    if (!user) {
      throw new Error('Invalid username or password');
    }
    if (user.password !== password) {
      throw new Error('Invalid username or password');
    }
    return user;
  }
}
